### ISSUE TITLE
Missed repaints while typing

### ISSUE THREAD

**Issue by Screwtapello** [mawww/kakoune#1804]
### Steps 

1. Run Kakoune with a cut-down version of my kakrc:

        kak -n -e 'set global static_words %sh{ tr "\n" ":" < /usr/share/dict/british-english-insane }; set global modelinefmt %{U+%sh{printf "%04x" "$kak_cursor_char_value"}}'

    On my Debian PC, `british-english-insane` is a list of ~650,000 words, which provides me with nice completion while writing English prose. The modification to the modeline shows the Unicode code-point under the cursor.
2. Start typing quickly. In my experience, the problem occurs most frequently with letter-pairs that are typed with opposite hands, so I can hit the two keys near-simultaneously—in particular, I've been testing with "of" and "an".

### Outcome 

The first character appears on-screen along with a list of completions beginning with that character. The second character does *not* appear, even if I wait seconds. If I type a third character, both the second and third characters immediately appear, along with an updated list of completions. If instead I hit backspace, no change appears on-screen but the phantom second character is deleted.

Basically, it's like Kakoune correctly updates its data model after each character, but fails to render the last update to the screen.

### Expected 

Kakoune should always update the screen within a fraction of a second of receiving a keypress

### Notes

I can't reproduce this with `british-english-small`, although I'm sure it's happened to me occasionally in real-world usage. I also can't reproduce if I use `seq 1 9999999` instead of a dictionary; apparently Kakoune can index that much more efficiently than English words. I also can't reproduce it with the standard modeline; apparently the shell launched for every key-press is one of the players in this race condition.
